Initial RM CI Pipelines

AdamHawtin opened the pull request:
### Motivation and Context
Pipeline for deploying the RM services to a CI environment in kubernetes. There are two functions of the pipeline, the first is to run through the kubernetes configuration files for the RM services, deploying them based on master in the census-rm-kubernetes repo. The second is to check for new builds of each RM service image latest tag in the container registry, then runs a `kubectl patch` with a timestamp to trigger kubernetes to pull the newly built image. 

The pipeline in running in the census concourse and has successfully deployed and updated the RM services.

As a stop gap measure the jobs currently have a hardcoded wait after running to give the services time to start up. This is limited in that it won't fail the concourse job if the deployment fails to start up, but it does prevent the deployments from running on top of each other which has caused issues. The aim is to replace this with dynamic waits which actually check the service availability which can then fail the job accordingly if the deployment does not start up in a given time.

### What has Changed
- New pipeline to deploy RM to a CI/dev environment
- New example secrets
- Updated README
- Updated gitignore

### Links
https://trello.com/c/pnfYl0em/435-concourse-13
https://github.com/ONSdigital/census-rm-terraform/pull/5

AdamHawtin committed:
First pass CI kubernetes apply pipeline

AdamHawtin committed:
First pass pipeline to force pull latest images in CI

AdamHawtin committed:
Organise file structure, update README and gitignore

AdamHawtin committed:
Don't pull images, naming changes, update gitignore

AdamHawtin commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Concourse (13)](https://trello.com/c/pnfYl0em/435-concourse-13)

AdamHawtin committed:
Trigger kube config jobs on commits to repo, use samplesvc stub in latest deploys

AdamHawtin committed:
Use an individual job for each services kubernetes files

AdamHawtin committed:
Update README

AdamHawtin committed:
Add image-digest to patched labels in latest deploy pipeline
Use existing sample service until the stub config is available

AdamHawtin committed:
Merge the two CI pipelines into one
This enables use of serial groups to stop deployments from conflicting and causing issues

davidmort commented on README.md:
```diff
@@ -1,2 +1,27 @@
+### CI Kubernetes Pipeline
+Specified in `pipelines/ci-kubernetes-pipeline.yml`, this pipeline has two functions.
+
+The first pulls from the `census-rm-kubernetes` repo triggered by commits to master and applies the latest microservice and handler configs to the kubernetes cluster specified by the pipeline configuration.
```
could make "census-rm-kubernetes" a link to the repo

davidmort commented on README.md:
```diff
@@ -1,2 +1,27 @@
+
+| Variable | Format | Description |
+| --- | --- | ---: |
+| kubernetes-server | https://\<host>:\<port> | The URL of the kubernetes server endpoint |
```
Might be worth mentioning "master" in the description, just to be specific.....

davidmort commented on README.md:
```diff
@@ -1,2 +1,27 @@
+| --- | --- | ---: |
+| kubernetes-server | https://\<host>:\<port> | The URL of the kubernetes server endpoint |
+| kubernetes-namespace | String | The kubernetes namespace for concourse to deploy to |
+| kubernetes-token | JWT String | A JWT access token. Must be for an account with the required roles in kubernetes |
```
might be worth being a bit more specific and mention "service account" not just "account" in the description

AdamHawtin committed:
Re-write pipeline to use gcloud IAM auth with scripts to run kubectl commands

AdamHawtin committed:
Add sleeps to wait for service start ups as a stop gap measure until we can do a dynamic check

kierenbeckett commented:
`kubectl rollout status deployment <deployment>`

AdamHawtin committed:
Task files for upcoming refactor

AdamHawtin committed:
Double quote selectors in tasks

AdamHawtin committed:
Move in line task definitions out to separate task files

AdamHawtin committed:
Use rollout status to wait on individual service startup

AdamHawtin committed:
Specify units for kubectl timeout

AdamHawtin committed:
Add namespace to rollout status command

AdamHawtin committed:
Add apply service and deploy task ahead of pipeline changes

AdamHawtin committed:
Fix mismatched param

AdamHawtin committed:
Use individual jobs to apply service and deploy specs

AdamHawtin committed:
Update README, update example secrets, remove unused task

Increase some timeouts as services aren't starting quick enough

Delete unused vars
